File: python3/bag_of_tokens.py
"""
Space   : O(n)
Time    : O(n log n)
"""

class Solution:
    def bagOfTokensScore(self, tokens: List[int], power: int) -> int:
        ans = 0
        
        tokens.sort()
        
        score = 0
        while tokens:
            while tokens and power >= tokens[0]:
                t = tokens.pop(0)
                power -= t
                score += 1
            
            ans = max(ans, score)
            
            if not tokens:
                break
                
            if score and power < tokens[0]:
                power += tokens.pop()
                score -= 1
            else:
                break
        
        return ans

File: golang/bag_of_tokens.go
package main

/*
Space	: O(n)
Time	: O(n log n)
*/

import "sort"

func bagOfTokensScore(tokens []int, power int) int {
	ans, score := 0, 0
	sort.Ints(tokens)

	for len(tokens) > 0 {
		for len(tokens) > 0 && power >= tokens[0] {
			t := tokens[0]
			tokens = tokens[1:]
			power -= t
			score++
		}

		ans = Max(ans, score)

		if len(tokens) == 0 {
			break
		}

		if score > 0 && power < tokens[0] {
			t := tokens[len(tokens)-1]
			tokens = tokens[:len(tokens)-1]
			power += t
			score -= 1
		} else {
			break
		}

	}

	return ans
}
